File: builtin-extensions/serverkit-email/backend/email.py
"""Email Server API endpoints for managing mail services, domains, accounts, and DNS.

Lives in the serverkit-email extension (app.plugins.serverkit-email.email after
install). Intra-extension modules are imported relatively; core services
(postfix relay, DNS, relay-config, RBAC) are imported absolutely — the mail
server extends the core, never the other way around.
"""
from flask import Blueprint, request, jsonify

from app.middleware.rbac import admin_required, viewer_required
from app.services.dns_provider_service import DNSProviderService
from app.services.postfix_service import PostfixService
from app.services.email_relay_service import EmailRelayService
from .email_service import EmailService
from .spamassassin_service import SpamAssassinService
from .roundcube_service import RoundcubeService

email_bp = Blueprint('email', __name__)


# ── Status & Installation ──

@email_bp.route('/status', methods=['GET'])
@viewer_required
def get_status():
    """Get aggregate email server status."""
    roundcube = RoundcubeService.get_status()
    status = EmailService.get_status()
    status['roundcube'] = roundcube
    return jsonify(status), 200


@email_bp.route('/install', methods=['POST'])
@admin_required
def install():
    """Install and configure all email components."""
    data = request.get_json() or {}
    hostname = data.get('hostname')
    result = EmailService.install_all(hostname)
    return jsonify(result), 200 if result.get('success') else 500


@email_bp.route('/service/<component>/<action>', methods=['POST'])
@admin_required
def control_service(component, action):
    """Start/stop/restart an email component."""
    result = EmailService.control_service(component, action)
    return jsonify(result), 200 if result.get('success') else 400


# ── Domains ──

@email_bp.route('/domains', methods=['GET'])
@viewer_required
def list_domains():
    """List all email domains."""
    domains = EmailService.get_domains()
    return jsonify({'domains': domains}), 200


@email_bp.route('/domains', methods=['POST'])
@admin_required
def add_domain():
    """Add an email domain."""
    data = request.get_json()
    if not data or not data.get('name'):
        return jsonify({'success': False, 'error': 'Domain name is required'}), 400
    result = EmailService.add_domain(
        data['name'],
        dns_provider_id=data.get('dns_provider_id'),
        dns_zone_id=data.get('dns_zone_id'),
    )
    return jsonify(result), 201 if result.get('success') else 400


@email_bp.route('/domains/<int:domain_id>', methods=['GET'])
@viewer_required
def get_domain(domain_id):
    """Get domain details."""
    result = EmailService.get_domain(domain_id)
    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 404


@email_bp.route('/domains/<int:domain_id>', methods=['DELETE'])
@admin_required
def remove_domain(domain_id):
    """Remove an email domain."""
    result = EmailService.remove_domain(domain_id)
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/domains/<int:domain_id>/verify-dns', methods=['POST'])
@viewer_required
def verify_dns(domain_id):
    """Verify DNS records for a domain."""
    result = EmailService.verify_dns(domain_id)
    return jsonify(result), 200


@email_bp.route('/domains/<int:domain_id>/deploy-dns', methods=['POST'])
@admin_required
def deploy_dns(domain_id):
    """Deploy DNS records via provider API."""
    from app.models.email import EmailDomain
    domain = EmailDomain.query.get(domain_id)
    if not domain:
        return jsonify({'success': False, 'error': 'Domain not found'}), 404
    if not domain.dns_provider_id or not domain.dns_zone_id:
        return jsonify({'success': False, 'error': 'No DNS provider configured for this domain'}), 400
    result = DNSProviderService.deploy_email_records(
        domain.dns_provider_id,
        domain.dns_zone_id,
        domain.name,
        domain.dkim_selector or 'default',
        domain.dkim_public_key or '',
    )
    return jsonify(result), 200 if result.get('success') else 400


# ── Accounts ──

@email_bp.route('/domains/<int:domain_id>/accounts', methods=['GET'])
@viewer_required
def list_accounts(domain_id):
    """List email accounts for a domain."""
    accounts = EmailService.get_accounts(domain_id)
    return jsonify({'accounts': accounts}), 200


@email_bp.route('/domains/<int:domain_id>/accounts', methods=['POST'])
@admin_required
def create_account(domain_id):
    """Create an email account."""
    data = request.get_json()
    if not data:
        return jsonify({'success': False, 'error': 'Request body required'}), 400
    username = data.get('username')
    password = data.get('password')
    if not username or not password:
        return jsonify({'success': False, 'error': 'Username and password are required'}), 400
    result = EmailService.add_account(
        domain_id,
        username,
        password,
        quota_mb=data.get('quota_mb', 1024),
    )
    return jsonify(result), 201 if result.get('success') else 400


@email_bp.route('/accounts/<int:account_id>', methods=['GET'])
@viewer_required
def get_account(account_id):
    """Get account details."""
    from app.models.email import EmailAccount
    account = EmailAccount.query.get(account_id)
    if not account:
        return jsonify({'success': False, 'error': 'Account not found'}), 404
    return jsonify({'success': True, 'account': account.to_dict()}), 200


@email_bp.route('/accounts/<int:account_id>', methods=['PUT'])
@admin_required
def update_account(account_id):
    """Update account settings."""
    data = request.get_json() or {}
    result = EmailService.update_account(
        account_id,
        quota_mb=data.get('quota_mb'),
        is_active=data.get('is_active'),
    )
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/accounts/<int:account_id>', methods=['DELETE'])
@admin_required
def delete_account(account_id):
    """Delete an email account."""
    result = EmailService.delete_account(account_id)
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/accounts/<int:account_id>/password', methods=['POST'])
@admin_required
def change_password(account_id):
    """Change account password."""
    data = request.get_json()
    if not data or not data.get('password'):
        return jsonify({'success': False, 'error': 'New password is required'}), 400
    result = EmailService.change_password(account_id, data['password'])
    return jsonify(result), 200 if result.get('success') else 400


# ── Aliases ──

@email_bp.route('/domains/<int:domain_id>/aliases', methods=['GET'])
@viewer_required
def list_aliases(domain_id):
    """List email aliases for a domain."""
    aliases = EmailService.get_aliases(domain_id)
    return jsonify({'aliases': aliases}), 200


@email_bp.route('/domains/<int:domain_id>/aliases', methods=['POST'])
@admin_required
def create_alias(domain_id):
    """Create an email alias."""
    data = request.get_json()
    if not data or not data.get('source') or not data.get('destination'):
        return jsonify({'success': False, 'error': 'Source and destination are required'}), 400
    result = EmailService.add_alias(domain_id, data['source'], data['destination'])
    return jsonify(result), 201 if result.get('success') else 400


@email_bp.route('/aliases/<int:alias_id>', methods=['DELETE'])
@admin_required
def delete_alias(alias_id):
    """Delete an email alias."""
    result = EmailService.remove_alias(alias_id)
    return jsonify(result), 200 if result.get('success') else 400


# ── Forwarding Rules ──

@email_bp.route('/accounts/<int:account_id>/forwarding', methods=['GET'])
@viewer_required
def list_forwarding(account_id):
    """List forwarding rules for an account."""
    rules = EmailService.get_forwarding(account_id)
    return jsonify({'rules': rules}), 200


@email_bp.route('/accounts/<int:account_id>/forwarding', methods=['POST'])
@admin_required
def create_forwarding(account_id):
    """Create a forwarding rule."""
    data = request.get_json()
    if not data or not data.get('destination'):
        return jsonify({'success': False, 'error': 'Destination is required'}), 400
    result = EmailService.add_forwarding(
        account_id,
        data['destination'],
        keep_copy=data.get('keep_copy', True),
    )
    return jsonify(result), 201 if result.get('success') else 400


@email_bp.route('/forwarding/<int:rule_id>', methods=['PUT'])
@admin_required
def update_forwarding(rule_id):
    """Update a forwarding rule."""
    data = request.get_json() or {}
    result = EmailService.update_forwarding(
        rule_id,
        destination=data.get('destination'),
        keep_copy=data.get('keep_copy'),
        is_active=data.get('is_active'),
    )
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/forwarding/<int:rule_id>', methods=['DELETE'])
@admin_required
def delete_forwarding(rule_id):
    """Delete a forwarding rule."""
    result = EmailService.remove_forwarding(rule_id)
    return jsonify(result), 200 if result.get('success') else 400


# ── DNS Providers ──

@email_bp.route('/dns-providers', methods=['GET'])
@viewer_required
def list_dns_providers():
    """List configured DNS providers."""
    providers = DNSProviderService.list_providers()
    return jsonify({'providers': providers}), 200


@email_bp.route('/dns-providers', methods=['POST'])
@admin_required
def add_dns_provider():
    """Add a DNS provider."""
    data = request.get_json()
    if not data or not data.get('name') or not data.get('provider') or not data.get('api_key'):
        return jsonify({'success': False, 'error': 'Name, provider, and api_key are required'}), 400
    result = DNSProviderService.add_provider(
        name=data['name'],
        provider=data['provider'],
        api_key=data['api_key'],
        api_secret=data.get('api_secret'),
        api_email=data.get('api_email'),
        is_default=data.get('is_default', False),
    )
    return jsonify(result), 201 if result.get('success') else 400


@email_bp.route('/dns-providers/<int:provider_id>', methods=['DELETE'])
@admin_required
def remove_dns_provider(provider_id):
    """Remove a DNS provider."""
    result = DNSProviderService.remove_provider(provider_id)
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/dns-providers/<int:provider_id>/test', methods=['POST'])
@admin_required
def test_dns_provider(provider_id):
    """Test DNS provider connection."""
    result = DNSProviderService.test_connection(provider_id)
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/dns-providers/<int:provider_id>/zones', methods=['GET'])
@viewer_required
def list_dns_zones(provider_id):
    """List DNS zones from a provider."""
    result = DNSProviderService.list_zones(provider_id)
    return jsonify(result), 200 if result.get('success') else 400


# ── Outbound SMTP Relay (smarthost) ──

@email_bp.route('/relay', methods=['GET'])
@viewer_required
def get_relay():
    """Get the outbound SMTP relay configuration (password masked)."""
    return jsonify(EmailRelayService.get_config()), 200


@email_bp.route('/relay', methods=['PUT'])
@admin_required
def save_relay():
    """Save the relay configuration and apply it to Postfix (if installed)."""
    data = request.get_json() or {}
    if not data.get('host'):
        return jsonify({'success': False, 'error': 'A relay host is required'}), 400
    result = EmailRelayService.save_config(data)
    return jsonify(result), 200


@email_bp.route('/relay/test', methods=['POST'])
@admin_required
def test_relay():
    """Open a real SMTP connection to validate the relay credentials."""
    data = request.get_json() or {}
    result = EmailRelayService.test(data)
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/relay', methods=['DELETE'])
@admin_required
def disable_relay():
    """Disable the outbound relay (revert Postfix to direct delivery)."""
    result = EmailRelayService.disable()
    return jsonify(result), 200


# ── SpamAssassin Config ──

@email_bp.route('/spam/config', methods=['GET'])
@viewer_required
def get_spam_config():
    """Get SpamAssassin configuration."""
    result = SpamAssassinService.get_config()
    return jsonify(result), 200


@email_bp.route('/spam/config', methods=['PUT'])
@admin_required
def update_spam_config():
    """Update SpamAssassin configuration."""
    data = request.get_json()
    if not data:
        return jsonify({'success': False, 'error': 'Request body required'}), 400
    result = SpamAssassinService.configure(data)
    if result.get('success'):
        SpamAssassinService.reload()
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/spam/update-rules', methods=['POST'])
@admin_required
def update_spam_rules():
    """Update SpamAssassin rules."""
    result = SpamAssassinService.update_rules()
    return jsonify(result), 200 if result.get('success') else 400


# ── Roundcube Webmail ──

@email_bp.route('/webmail/status', methods=['GET'])
@viewer_required
def webmail_status():
    """Get Roundcube webmail status."""
    result = RoundcubeService.get_status()
    return jsonify(result), 200


@email_bp.route('/webmail/install', methods=['POST'])
@admin_required
def webmail_install():
    """Install Roundcube webmail."""
    data = request.get_json() or {}
    result = RoundcubeService.install(
        imap_host=data.get('imap_host', 'host.docker.internal'),
        smtp_host=data.get('smtp_host', 'host.docker.internal'),
        domain=(data.get('domain') or '').strip() or None,
    )
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/webmail/service/<action>', methods=['POST'])
@admin_required
def webmail_control(action):
    """Start/stop/restart Roundcube."""
    actions = {
        'start': RoundcubeService.start,
        'stop': RoundcubeService.stop,
        'restart': RoundcubeService.restart,
    }
    if action not in actions:
        return jsonify({'success': False, 'error': 'Invalid action'}), 400
    result = actions[action]()
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/webmail/configure-proxy', methods=['POST'])
@admin_required
def webmail_configure_proxy():
    """Configure Nginx reverse proxy for Roundcube."""
    data = request.get_json()
    if not data or not data.get('domain'):
        return jsonify({'success': False, 'error': 'Domain is required'}), 400
    result = RoundcubeService.configure_nginx_proxy(data['domain'])
    return jsonify(result), 200 if result.get('success') else 400


# ── Mail Queue & Logs ──

@email_bp.route('/queue', methods=['GET'])
@viewer_required
def get_queue():
    """Get Postfix mail queue."""
    result = PostfixService.get_queue()
    return jsonify(result), 200


@email_bp.route('/queue/flush', methods=['POST'])
@admin_required
def flush_queue():
    """Flush the mail queue."""
    result = PostfixService.flush_queue()
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/queue/<queue_id>', methods=['DELETE'])
@admin_required
def delete_queue_item(queue_id):
    """Delete a message from the queue."""
    result = PostfixService.delete_from_queue(queue_id)
    return jsonify(result), 200 if result.get('success') else 400


@email_bp.route('/logs', methods=['GET'])
@viewer_required
def get_logs():
    """Get mail logs."""
    lines = request.args.get('lines', 100, type=int)
    result = PostfixService.get_logs(lines)
    return jsonify(result), 200

File: backend-rs/crates/sk-server/src/routes/email.rs
use crate::error::{ApiError, ApiResult};
use crate::extract::AuthUser;
use crate::state::SharedState;
use axum::extract::{Path, Query, State};
use axum::routing::{delete, get, post, put};
use axum::{Json, Router};
use serde::Deserialize;
use serde_json::{json, Value};
fn internal(e: anyhow::Error) -> ApiError {
    e.into()
}
fn body(b: Option<Json<Value>>) -> Value {
    match b.map(|x| x.0).unwrap_or_else(|| json!({})) {
        Value::String(s) => serde_json::from_str(&s).unwrap_or_else(|_| json!({"value":s})),
        v => v,
    }
}
#[derive(Deserialize)]
struct Lines {
    lines: Option<i64>,
}
pub fn router() -> Router<SharedState> {
    Router::new()
        .route("/status", get(status))
        .route("/install", post(install))
        .route("/service/{component}/{action}", post(control))
        .route("/domains", get(domains).post(add_domain))
        .route("/domains/{id}", get(domain).delete(delete_domain))
        .route("/domains/{id}/verify-dns", post(verify_dns))
        .route("/domains/{id}/deploy-dns", post(deploy_dns))
        .route("/domains/{id}/accounts", get(accounts).post(create_account))
        .route(
            "/accounts/{id}",
            get(account).put(update_account).delete(delete_account),
        )
        .route("/accounts/{id}/password", post(change_password))
        .route("/domains/{id}/aliases", get(aliases).post(create_alias))
        .route("/aliases/{id}", delete(delete_alias))
        .route(
            "/accounts/{id}/forwarding",
            get(forwarding).post(create_forwarding),
        )
        .route(
            "/forwarding/{id}",
            put(update_forwarding).delete(delete_forwarding),
        )
        .route("/dns-providers", get(providers).post(add_provider))
        .route("/dns-providers/{id}", delete(delete_provider))
        .route("/dns-providers/{id}/test", post(test_provider))
        .route("/dns-providers/{id}/zones", get(zones))
        .route("/relay", get(relay).put(update_relay).delete(disable_relay))
        .route("/relay/test", post(test_relay))
        .route("/spam/config", get(spam).put(update_spam))
        .route("/spam/update-rules", post(update_spam_rules))
        .route("/webmail/status", get(webmail_status))
        .route("/webmail/install", post(webmail_install))
        .route("/webmail/service/{action}", post(webmail_control))
        .route("/webmail/configure-proxy", post(webmail_proxy))
        .route("/queue", get(queue))
        .route("/queue/flush", post(flush_queue))
        .route("/queue/{id}", delete(delete_queue))
        .route("/logs", get(logs))
}
async fn status(State(s): State<SharedState>, AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::status(&s.db).await.map_err(internal)?))
}
async fn install(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::install(&s.db, &body(b)).await.map_err(internal)?,
    ))
}
async fn control(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path((component, action)): Path<(String, String)>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::control(&s.db, &component, &action)
            .await
            .map_err(internal)?,
    ))
}
async fn domains(State(s): State<SharedState>, AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::domains(&s.db).await.map_err(internal)?))
}
async fn add_domain(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::add_domain(&s.db, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn domain(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::domain(&s.db, &id).await.map_err(internal)?))
}
async fn delete_domain(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::delete_domain(&s.db, &id)
            .await
            .map_err(internal)?,
    ))
}
async fn verify_dns(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::verify_dns(&s.db, &id).await.map_err(internal)?,
    ))
}
async fn deploy_dns(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::deploy_dns(&s.db, &id).await.map_err(internal)?,
    ))
}
async fn accounts(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::accounts(&s.db, &id).await.map_err(internal)?,
    ))
}
async fn create_account(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::create_account(&s.db, &id, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn account(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::account(&s.db, &id).await.map_err(internal)?))
}
async fn update_account(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::update_account(&s.db, &id, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn delete_account(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::delete_account(&s.db, &id)
            .await
            .map_err(internal)?,
    ))
}
async fn change_password(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::change_password(&s.db, &id, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn aliases(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::aliases(&s.db, &id).await.map_err(internal)?))
}
async fn create_alias(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::create_alias(&s.db, &id, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn delete_alias(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::delete_alias(&s.db, &id).await.map_err(internal)?,
    ))
}
async fn forwarding(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::forwarding(&s.db, &id).await.map_err(internal)?,
    ))
}
async fn create_forwarding(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::create_forwarding(&s.db, &id, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn update_forwarding(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::update_forwarding(&s.db, &id, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn delete_forwarding(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::delete_forwarding(&s.db, &id)
            .await
            .map_err(internal)?,
    ))
}
async fn providers(State(s): State<SharedState>, AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::providers(&s.db).await.map_err(internal)?))
}
async fn add_provider(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::add_provider(&s.db, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn delete_provider(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::delete_provider(&s.db, &id)
            .await
            .map_err(internal)?,
    ))
}
async fn test_provider(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::test_provider(&s.db, &id)
            .await
            .map_err(internal)?,
    ))
}
async fn zones(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::zones(&s.db, &id).await.map_err(internal)?))
}
async fn relay(State(s): State<SharedState>, AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::relay(&s.db).await.map_err(internal)?))
}
async fn update_relay(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::update_relay(&s.db, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn disable_relay(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::disable_relay(&s.db).await.map_err(internal)?,
    ))
}
async fn test_relay(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::test_relay(&s.db, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn spam(State(s): State<SharedState>, AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::spam(&s.db).await.map_err(internal)?))
}
async fn update_spam(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::update_spam(&s.db, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn update_spam_rules(AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::update_spam_rules().await.map_err(internal)?))
}
async fn webmail_status(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::webmail_status(&s.db).await.map_err(internal)?,
    ))
}
async fn webmail_install(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::webmail_install(&s.db).await.map_err(internal)?,
    ))
}
async fn webmail_control(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(action): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::webmail_control(&s.db, &action)
            .await
            .map_err(internal)?,
    ))
}
async fn webmail_proxy(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::webmail_proxy(&s.db, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn queue(State(s): State<SharedState>, AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::queue(&s.db).await.map_err(internal)?))
}
async fn flush_queue(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
) -> ApiResult<Json<Value>> {
    Ok(Json(sk_email::flush_queue(&s.db).await.map_err(internal)?))
}
async fn delete_queue(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::delete_queue(&s.db, &id).await.map_err(internal)?,
    ))
}
async fn logs(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Query(q): Query<Lines>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_email::logs(&s.db, q.lines.unwrap_or(100))
            .await
            .map_err(internal)?,
    ))
}
